qt/RpcQueue.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * This file Copyright (C) 2015 Mnemosyne LLC
+ *
+ * It may be used under the GNU GPL versions 2 or 3
+ * or any future license endorsed by Mnemosyne LLC.
+ *
+ * $Id$
+ */
+
+#ifndef QTR_RPC_QUEUE_H
+#define QTR_RPC_QUEUE_H
+
+#include <functional>
+
+#include <QFutureWatcher>
+#include <QQueue>
+
+#include "RpcClient.h"
+
+class RpcQueue: public QObject
+{
+  Q_OBJECT
+
+  /*
+   * Internally queued function. Takes the last response future, makes a
+   * request and returns a new response future.
+   */
+  typedef std::function<QFuture<RpcResponse> (QFuture<RpcResponse>)> InternalQueuedFunction;
+
+  /*
+   * Standard queued function. Takes the last response, makes a request and
+   * returns a response future corresponding to the newly made request.
+   */
+  typedef std::function<QFuture<RpcResponse> (RpcResponse)> QueuedFunction;
+
+  /*
+   * Queued function without a result. Takes the last response, performs
+   * arbitrary actions and returns nothing.
+   */
+  typedef std::function<void (RpcResponse)> QueuedFunctionNoResult;
+
+  /*
+   * Internally stored error handler function. Takes the last response future and returns nothing.
+   */
+  typedef std::function<void (QFuture<RpcResponse>)> InternalErrorHandlerFunction;
+
+  /*
+   * Error handler function. Takes the last response and returns nothing.
+   */
+  typedef std::function<void (RpcResponse)> ErrorHandlerFunction;
+
+public:
+  explicit
+  RpcQueue (QObject * parent = 0);
+
+  void
+  setTolerateErrors (bool tolerateErrors = true) { myTolerateErrors = tolerateErrors; }
+
+  void
+  add (QueuedFunction func, ErrorHandlerFunction onError = ErrorHandlerFunction ());
+
+  void
+  add_aux (QueuedFunctionNoResult func);
```

Comment:
The intent is that adds an "auxiliary" processing function which does not introduce its own request (ultimately, the next queued function gets passed the same response).
What would you suggest?